### PR TITLE
pyvmomi backed vmware module util

### PR DESCRIPTION
**Pull request by jctanner**
##### ISSUE TYPE
- Feature Pull Request
##### ANSIBLE VERSION

```
2.2
```
##### SUMMARY

A revised module util for vmware that is backed by pyvmomi instead of psphere/pysphere.

### CODE REVIEW

File: lib/ansible/module_utils/vmware_pyvmomi.py [hunk 1]
```diff
@@ -0,0 +1,688 @@
...
+    from pyVim.connect import SmartConnect, Disconnect
+    HAS_PYVMOMI = True
+except ImportError:
+    pass
+
+import atexit
+import os
+import ssl
+import time
+from pprint import pprint
```
Comment (alikins) [inline lib/ansible/module_utils/vmware_pyvmomi.py:40]: unused import

File: lib/ansible/module_utils/vmware_pyvmomi.py [hunk 2]
```diff
@@ -0,0 +1,688 @@
...
+            err = get_exception()
+            self.module.fail_json(msg="Cannot connect to %s: %s" %
+                             (kwargs['host'], err))
+        atexit.register(Disconnect, self.si)
+        self.content = self.si.RetrieveContent()
+
+    def _build_folder_tree(self, folder, tree={}, treepath=None):
+
+        tree = {'virtualmachines': [],
+                        'subfolders': {},
```
Comment (alikins) [inline lib/ansible/module_utils/vmware_pyvmomi.py:82]: fix indent

File: lib/ansible/module_utils/vmware_pyvmomi.py [hunk 3]
```diff
@@ -0,0 +1,688 @@
...
+        if hasattr(ssl, 'SSLContext'):
+            context = ssl.SSLContext(ssl.PROTOCOL_TLSv1)
+            context.verify_mode = ssl.CERT_NONE
+            kwargs['sslContext'] = context
+
+        # CONNECT TO THE SERVER
+        try:
+            self.si = SmartConnect(**kwargs)
+        except Exception:
+            err = get_exception()
```
Comment (alikins) [inline lib/ansible/module_utils/vmware_pyvmomi.py:73]: get_exception() isnt imported anywhere

File: lib/ansible/module_utils/vmware_pyvmomi.py [hunk 4]
```diff
@@ -0,0 +1,688 @@
...
+
+
+    def _build_folder_map(self, folder, vmap={}, inpath='/'):
+
+        ''' Build a searchable index for vms+uuids+folders '''
+
+        if type(folder) == tuple:
+            folder = folder[1]
+
+        if not 'names' in vmap:
```
Comment (alikins) [inline lib/ansible/module_utils/vmware_pyvmomi.py:112]: pep8 stuff

```
if 'names' not in vmap:
```

File: lib/ansible/module_utils/vmware_pyvmomi.py [hunk 5]
```diff
@@ -0,0 +1,688 @@
...
+        try:
+            self.si = SmartConnect(**kwargs)
+        except Exception:
+            err = get_exception()
+            self.module.fail_json(msg="Cannot connect to %s: %s" %
+                             (kwargs['host'], err))
+        atexit.register(Disconnect, self.si)
+        self.content = self.si.RetrieveContent()
+
+    def _build_folder_tree(self, folder, tree={}, treepath=None):
```
Comment (alikins) [inline lib/ansible/module_utils/vmware_pyvmomi.py:79]: Is 'tree' given a default mutable arg here intentionally? If so, that would be worth a comment explaining why.

If not then...

``` python
def _build_folder_tree(self, folder, tree=None, treepath=None):
    tree = tree or {'virtualmachines': [],
                             'subfolders': {},
                             'name': folder.name}

```

ditto for 'vmap' on _build_folder_map below.

File: lib/ansible/module_utils/vmware_pyvmomi.py [hunk 6]
```diff
@@ -0,0 +1,688 @@
...
+
+        else:
+            if firstmatch:
+                vm = get_obj(self.content, [vim.VirtualMachine], name)
+            else:
+                matches = []
+                vmList = get_all_objs(self.content, [vim.VirtualMachine])
+                for thisvm in vmList:
+                    if thisvm.config == None:
+                        import epdb; epdb.st()
```
Comment (alikins) [inline lib/ansible/module_utils/vmware_pyvmomi.py:210]: rm debugger hook.

File: lib/ansible/module_utils/vmware_pyvmomi.py [hunk 7]
```diff
@@ -0,0 +1,688 @@
...
+	"""
+	Set the power status for a VM determined by the current and
+	requested states. force is forceful
+	"""
+        facts = self.gather_facts(vm)
+        expected_state = state.replace('_', '').lower()
+        current_state = facts['hw_power_status'].lower()
+        result = {}
+
+	# Need Force
```
Comment (alikins) [inline lib/ansible/module_utils/vmware_pyvmomi.py:232]: lots of mixed space/tabs through here.

File: lib/ansible/module_utils/vmware_pyvmomi.py [hunk 8]
```diff
@@ -0,0 +1,688 @@
...
+                elif expected_state == 'restarted':
+                    if current_state in ('poweredon', 'poweringon', 'resetting'):
+                        task = vm.Reset()
+                    else:
+                        result = {'changed': False, 'failed': True, 
+                                  'msg': "Cannot restart VM in the current state %s" % current_state}
+
+            except Exception:
+                result = {'changed': False, 'failed': True, 
+                          'msg': get_exception()}
```
Comment (alikins) [inline lib/ansible/module_utils/vmware_pyvmomi.py:261]: missing import of get_exception()

File: lib/ansible/module_utils/vmware_pyvmomi.py [hunk 9]
```diff
@@ -0,0 +1,688 @@
...
+            if len(path) == 1:
+                return o
+            else:
+                return _find_path_in_tree(o['subfolders'], path[1:])
+
+    return None
+
+
+def _get_folderid_for_path(vsphere_client, datacenter, path):
+    content = vsphere_client._retrieve_properties_traversal(property_names=['name', 'parent'], obj_type=MORTypes.Folder)
```
Comment (alikins) [inline lib/ansible/module_utils/vmware_pyvmomi.py:671]: MORTypes is not imported or defined.

File: lib/ansible/module_utils/vmware_pyvmomi.py (hunk 4 above)
Comment (bcoca) [inline lib/ansible/module_utils/vmware_pyvmomi.py:112]: this is one of the pep8 rules we don't need to follow

**Pull request closed without merging**